control-plane/agents/src/bin/core/controller/reconciler/volume/garbage_collector.rs

```diff
@@ -250,6 +258,56 @@ async fn disown_unused_replicas(
     GarbageCollector::squash_results(results)
 }
 
+/// Volume replicas could have been possibly expanded as part of volume resize,
+/// but the volume expansion failed overall leaving behind expanded replicas. So we
+/// bring back the replica size to what is expected by the volume spec.
+#[tracing::instrument(level = "debug", skip(context, volume), fields(volume.uuid = %volume.uuid(), request.reconcile = true))]
+async fn revert_expanded_replicas(
+    volume: &mut OperationGuardArc<VolumeSpec>,
+    context: &PollContext,
+) -> PollResult {
+    let target = context.specs().volume_target_nexus_rsc(volume.as_ref());
+    if let Some(tgt) = target {
+        let nexus_size = tgt.lock().size;
+        debug_assert_eq!(nexus_size, volume.as_ref().size);
+        if nexus_size != volume.as_ref().size {
+            // This should never happen, but just emit a warning.
+            volume.warn_span(|| tracing::info!(volume_size = %volume.as_ref().size, nexus_size = %nexus_size, "Nexus and Volume size mismatch!"));
+        }
+        // If there is a target present, we don't try to reclaim space from replicas because the
+        // attempt to downsize replica/lvol will potentially fail with -EBUSY if the lvol bdev
+        // has open descriptors.
+        return PollResult::Ok(PollerState::Idle);
+    }
+
+    for replica in context.specs().volume_replicas(volume.uuid()) {
+        let mut replica = match replica.operation_guard() {
+            Ok(guard) => guard,
+            Err(_) => continue,
+        };
+
+        if replica.as_ref().size != volume.as_ref().size {
```

Comment:
Should this be:
```suggestion
        if replica.as_ref().size > volume.as_ref().size {
```
Otherwise it could potentially grow rather than shrink?

Reply:
> Otherwise it could potentially grow rather than shrink?

Changed to `>` . PTAL. Theoretically yes it would mistakenlygrow, but I don't see a case that can put us in such situation.

Reply:
hmm what if an "extra" replica was offline so was not resized up as part of the volume resize?
although this is interesting, if this can happen I wonder if we need to expand a replica when we want to add it to the nexus but it's not the correct size?

Reply:
> hmm what if an "extra" replica was offline so was not resized up as part of the volume resize?

If a replica that's part of a volume is offline, then the pre-check fails and volume resize doesn't initiate at all.

Reply:
So if a node goes offline volume resizes become blocked?

Reply:
> So if a node goes offline volume resizes become blocked?

Not blocked I meant, it is rejected with failed precondition(from the scheduling pre-check that we do) saying 
> Replicas  can't be resized due to required capacity ({}) or state(Online) not met.

Reply:
I mean if a volume replica node is offline then you can't resize until you bring the node online? Even if the replica is not part of the nexus?